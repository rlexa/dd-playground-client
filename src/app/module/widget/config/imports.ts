import { CommonModule } from '@angular/common';
import { MatCardModule, MatListModule } from '@angular/material';
import { RouterModule } from '@angular/router';

export const imports = [CommonModule, RouterModule, MatCardModule, MatListModule];
