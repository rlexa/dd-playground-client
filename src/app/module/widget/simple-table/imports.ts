import { CommonModule } from '@angular/common';
import { MatFormFieldModule, MatInputModule, MatPaginatorModule, MatSortModule, MatTableModule } from '@angular/material';
import { RipupperPipeModule } from 'app/module/pipe/ripupper';
import { StartuppercasePipeModule } from 'app/module/pipe/startuppercase';
import { FlexboxModule } from 'dd-flexbox';

export const imports = [
  CommonModule,
  MatFormFieldModule, MatInputModule, MatPaginatorModule, MatSortModule, MatTableModule, FlexboxModule,
  RipupperPipeModule, StartuppercasePipeModule,
];
