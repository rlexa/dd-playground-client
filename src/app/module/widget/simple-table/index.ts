export * from './simple-table.component';
export * from './simple-table.module';
