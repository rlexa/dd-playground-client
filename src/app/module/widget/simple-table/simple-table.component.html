<ng-container *ngIf="columns$ | async as columns">
  <ng-container *ngIf="columns.length > 0">

    <mat-form-field class="childhv">
      <input matInput [value]="filter$ | async" (change)="setFilter($event.target.value)" (keyup)="setFilter($event.target.value)"
        placeholder="Filter">
    </mat-form-field>

    <div ddFlexitem="1 1" class="scroll-y-always">
      <table mat-table class="text-nowrap" [trackBy]="trackByIndex" [dataSource]="dataBatch$" matSort [matSortDisableClear]="true"
        [matSortDirection]="(sortAsc$ | async) ? 'asc' : 'desc'" [matSortActive]="sortBy$ | async" (matSortChange)="onSortChange($event.active, $event.direction !== 'desc')">

        <ng-container *ngFor="let col of columns; trackBy: trackByIndex;" [matColumnDef]="col">
          <th mat-header-cell *matHeaderCellDef [mat-sort-header]="col">{{ col | startuppercase | ripupper }}</th>
          <td mat-cell *matCellDef="let ii">
            <div class="break-text contenthv" [style.fontSize]="'1.2rem'">
              <span>
                {{ ii[col] }}
              </span>
            </div>
          </td>
        </ng-container>

        <tr mat-header-row *matHeaderRowDef="columns; sticky: true;"></tr>
        <tr mat-row *matRowDef="let row; columns: columns;" class="hoverable" (dblclick)="onRowDoubleClick(row)"></tr>
      </table>
    </div>

    <mat-paginator [hidePageSize]="true" [pageSize]="pageSize$ | async" [showFirstLastButtons]="true" [length]="total$ | async" [pageIndex]="page$ | async"
      (page)="setPage($event.pageIndex)"></mat-paginator>

  </ng-container>
</ng-container>