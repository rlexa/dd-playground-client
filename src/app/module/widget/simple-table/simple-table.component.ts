import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { FORMAT_DATE_TIMESTAMP } from 'app/presets';
import { DoneSubject, rxComplete, rxNext_, rxNull, rxTrue } from 'app/rx';
import { trackByIndex } from 'app/util';
import { BehaviorSubject, combineLatest, merge } from 'rxjs';
import { debounceTime, filter, map, takeUntil } from 'rxjs/operators';

@Component({
  selector: 'app-simple-table',
  templateUrl: './simple-table.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SimpleTableComponent implements OnDestroy, OnInit {
  @HostBinding('style.display') readonly styleDisplay = 'flex';
  @HostBinding('style.flexDirection') readonly styleFlexDirection = 'column';

  private readonly done$ = new DoneSubject();
  private readonly data$ = new BehaviorSubject([]);

  readonly FORMAT_DATE_TIMESTAMP = FORMAT_DATE_TIMESTAMP;

  readonly filter$ = new BehaviorSubject(<string>null);
  readonly sortAsc$ = new BehaviorSubject(true);
  readonly sortBy$ = new BehaviorSubject(<string>null);
  readonly page$ = new BehaviorSubject(0);
  readonly pageSize$ = new BehaviorSubject(10);

  readonly columns$ = this.data$.pipe(map(_ => Object.entries((_ || [])[0] || {}).filter(([key, val]) => typeof val !== 'object').map(([key, val]) => key)));
  private readonly dataFiltered$ = combineLatest(this.data$, this.columns$, this.filter$).pipe(
    debounceTime(0),
    map(([data, columns, filtering]) => (data || [])
      .filter(item => !filtering || columns.some(col => (item[col] || '').toString().toLocaleLowerCase().includes(filtering.toLocaleLowerCase())))),
    takeUntil(this.done$));
  readonly total$ = this.dataFiltered$.pipe(map(_ => (_ || []).length));

  readonly dataBatch$ = combineLatest(this.page$, this.pageSize$, this.sortAsc$, this.sortBy$, this.dataFiltered$)
    .pipe(
      debounceTime(0),
      map(([page, size, sortAsc, sortBy, data]) => (data || [])
        .sort((aa, bb) => {
          return sortBy ? (typeof aa[sortBy] === 'string' ? (aa[sortBy] as string).localeCompare(bb[sortBy]) : aa[sortBy] - bb[sortBy]) * (sortAsc ? 1 : -1) : 0;
        })
        .slice(page * size, page * size + size)),
      takeUntil(this.done$));

  setFilter = rxNext_(this.filter$);
  setPage = rxNext_(this.page$);
  setSortAsc = rxNext_(this.sortAsc$);
  setSortBy = rxNext_(this.sortBy$);

  trackByIndex = trackByIndex;

  @Output() dblClick = new EventEmitter<any>();

  @Input() set data(val: any[]) { this.data$.next(val); }

  ngOnDestroy() {
    this.done$.done();
    rxComplete(
      this.data$,
      this.filter$,
      this.page$,
      this.pageSize$,
      this.sortAsc$,
      this.sortBy$,
    );
  }

  ngOnInit() {
    merge(this.dataFiltered$, this.pageSize$, this.filter$, this.sortAsc$, this.sortBy$)
      .pipe(map(() => 0), filter(_ => _ !== this.page$.value), takeUntil(this.done$))
      .subscribe(rxNext_(this.page$));

    combineLatest(this.columns$, this.sortBy$)
      .pipe(filter(([columns, sortBy]) => !!sortBy && !columns.includes(sortBy)), takeUntil(this.done$))
      .subscribe(() => {
        rxNull(this.sortBy$);
        rxTrue(this.sortAsc$);
      });
  }

  onRowDoubleClick = (item: any) => this.dblClick.emit(item);

  onSortChange = (sortBy: string, sortAsc: boolean) => {
    this.setSortBy(sortBy);
    this.setSortAsc(sortAsc);
  }
}
