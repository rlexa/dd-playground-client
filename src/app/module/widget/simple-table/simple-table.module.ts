import {CommonModule} from '@angular/common';
import {NgModule} from '@angular/core';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatInputModule} from '@angular/material/input';
import {MatPaginatorModule} from '@angular/material/paginator';
import {MatSortModule} from '@angular/material/sort';
import {MatTableModule} from '@angular/material/table';
import {FlexboxModule} from '../../directive/flexbox';
import {RipupperPipeModule} from '../../pipe/ripupper';
import {StartuppercasePipeModule} from '../../pipe/startuppercase';
import {SimpleTableComponent} from './simple-table.component';

@NgModule({
  imports: [
    CommonModule,
    MatFormFieldModule,
    MatInputModule,
    MatPaginatorModule,
    MatSortModule,
    MatTableModule,
    FlexboxModule,
    RipupperPipeModule,
    StartuppercasePipeModule,
  ],
  exports: [SimpleTableComponent],
  declarations: [SimpleTableComponent],
})
class SimpleTableModule {}

export {SimpleTableModule, SimpleTableComponent};
