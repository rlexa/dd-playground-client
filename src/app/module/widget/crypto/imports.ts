import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import {
  MatButtonModule,
  MatCardModule,
  MatFormFieldModule,
  MatInputModule,
  MatListModule,
} from '@angular/material';
import { RouterModule } from '@angular/router';
import { CryptoApiModule } from 'app/module/service/crypto-api';
import { SimpleViewModule } from 'app/module/widget/simple-view';
import { FlexboxModule } from 'app/module/directive/flexbox';

export const imports = [
  CommonModule,
  FormsModule,
  RouterModule,
  MatButtonModule,
  MatCardModule,
  MatFormFieldModule,
  MatInputModule,
  MatListModule,
  FlexboxModule,
  CryptoApiModule,
  SimpleViewModule,
];
