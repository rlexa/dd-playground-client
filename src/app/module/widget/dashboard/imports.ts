import { CommonModule } from '@angular/common';
import { MatButtonModule, MatToolbarModule, MatTooltipModule } from '@angular/material';
import { RouterModule } from '@angular/router';
import { IconPipeModule } from 'app/module/pipe/icon';
import { RoutedContentModule } from 'app/module/widget/routed-content';
import { VersionModule } from 'app/module/widget/version';
import { WidgetsModule } from 'app/widgets/widgets.module';
import { FlexboxModule } from 'dd-flexbox';
import 'hammerjs';

export const imports = [
  CommonModule, RouterModule,
  MatButtonModule, MatToolbarModule, MatTooltipModule, FlexboxModule,
  IconPipeModule, VersionModule,
  // remove
  RoutedContentModule, WidgetsModule,
];
