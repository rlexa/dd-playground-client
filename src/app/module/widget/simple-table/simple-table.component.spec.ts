import {async, TestBed} from '@angular/core/testing';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {imports} from './imports';
import {SimpleTableComponent} from './simple-table.component';

describe('SimpleTableComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, ...imports],
      declarations: [SimpleTableComponent],
      providers: [],
    }).compileComponents();
  }));

  test('is created', () => {
    const fixture = TestBed.createComponent(SimpleTableComponent);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
