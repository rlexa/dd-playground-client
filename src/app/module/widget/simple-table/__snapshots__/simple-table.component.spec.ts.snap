// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTableComponent is created 1`] = `
<app-simple-table
  FORMAT_DATE_TIMESTAMP={[Function String]}
  columns$={[Function AnonymousSubject]}
  data$={[Function BehaviorSubject]}
  dataBatch$={[Function Observable]}
  dataFiltered$={[Function Observable]}
  dblClick={[Function EventEmitter]}
  done$={[Function DoneSubject]}
  filter$={[Function BehaviorSubject]}
  onRowDoubleClick={[Function Function]}
  onSortChange={[Function Function]}
  page$={[Function BehaviorSubject]}
  pageSize$={[Function BehaviorSubject]}
  setFilter={[Function Function]}
  setPage={[Function Function]}
  setSortAsc={[Function Function]}
  setSortBy={[Function Function]}
  sortAsc$={[Function BehaviorSubject]}
  sortBy$={[Function BehaviorSubject]}
  styleDisplay={[Function String]}
  styleFlexDirection={[Function String]}
  total$={[Function Observable]}
  trackByIndex={[Function Function]}
>
  
</app-simple-table>
`;
